### ISSUE TITLE
RFC: Release targets

### ISSUE THREAD

**Issue by dalehamel** [iovisor/bpftrace#1051]
This is a fork off of the discussion from #342 @danobi 

The idea here is to use the existing CI build for bpftrace to generate various release targets when the release process is initiated by `git tag`. 

These artifacts should be attached to a Release object from the github API transparently. 

It may also make sense to use this process to drive publishing artifacts to other systems, such as:

- Publishing a debian package to an iovisor-hosted apt repo
- Publishing a flatpak or snapcraft image
- Pushing a docker image to quay.io/iovisor/bpftrace or dockerhub

I think that this wolud help to make bpftrace more accessible.

This issue mostly supposes that #1041 makes sense, and bits like authenticating with quay.io/iovisor images would help to speed up CI runs, which is itself a concern of #1041 

# Implementation

Most of this comes down to someone with credentials on the various package publishing system to encrypt them and store them in the travis.yml

Travis itself provides deploy hooks that are probably flexible enough to initiate almost any process.

Most of these artifacts already exist, and deploy steps just need to be configured to copy them or do some packaging post-processing.

# Side benefit - Faster CI builds

If we can get access to push images from previous runs, we only rebuild layers when necessary. Each build type can have their own docker tag based on their ENV vars. This could help drop CI times significantly, depending on the docker pull speed.

**Comment by dalehamel** [comment]
Sample release process for travis: 

#before_deploy:
#  - FIXME copy binaries out of static builds
# deploy:
#   provider: releases
#   api_key:
#     secure: YOUR_API_KEY_ENCRYPTED
#   file:
#     - "FILE 1"
#     - "FILE 2"
#   skip_cleanup: true
#   on:
#     tags: true
#     condition: $RELEASE_BINARY = ON

**Comment by dalehamel** [comment]
@mmarchini @fbs @danobi and i discussed using github actions https://help.github.com/en/actions/automating-your-workflow-with-github-actions/configuring-workflows

The syntax is here https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions and doesn't seem too different than Travis.

I don't have access to set this up for iovisor/bpftrace, but perhaps I can explore it on my own fork and submit a config file analogous to the existing travis file?

Who would have access to set this up / could review the PR?

**Comment by dalehamel** [comment]
I've played around with github actions, they are incredible.

My only gripe is the caching limit is 2GB, but the amount of parallelism is amazing - 20 jobs up to 6 hours runtime each, so it makes up for it.

I think it can probably also just run as a non-required check pretty easily for other jobs.

The existing checks from travis could certainly be ported over as well.

For publishing releases, someone would just need to add a bot token to this repo:

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token

Which could automate the release process. Likewise, bot tokens for other distribution places like docker registrys, etc.

**Comment by mmarchini** [comment]
GitHub Actions are setup on repositories by default, so if you send a Pull Request with configs it will work just fine :)

> For publishing releases, someone would just need to add a bot token to this repo:

`GITHUB_TOKEN` is an environment variable which is available by default in a job. You can use it to interact with the GitHub API (upload artifacts, make comments, etc.). The only caveat is that `GITHUB_ACTIONS` will give read-only access to the GitHub API in jobs triggered from Pull Requests coming from forks, but in most cases that's fine.

**Comment by dalehamel** [comment]
> GitHub Actions are setup on repositories by default, so if you send a Pull Request with configs it will work just fine :)

Yeah it looks like it runs against my own fork, which makes a lot of sense.

> GITHUB_TOKEN is an environment variable which is available by default in a job. You can use it to interact with the GitHub API (upload artifacts, make comments, etc.). The only caveat is that GITHUB_ACTIONS will give read-only access to the GitHub API in jobs triggered from Pull Requests coming from forks, but in most cases that's fine.

I think that's definitely preferable, we only want tagged releases publishing artifacts I think. By default you can upload per-build artifacts pretty easily by just specifying it on the job. There is a separate action for uploading it to a release. I think automating the release process to be triggered by a tag is ideal. I think that if we can just get the necessary release secrets into the Secrets tab, they should be passed like on travis. Who has access to iovisor quay account to push images on master / tagged releases? Whatever secret is being used to message to IRC can probably added likewise, though don't want to be too spammy there. 

It also looks like by default checks don't have to block a pull request. Repo maintainers can go in and mark checks as non-blocking to merge to master. I think that this should be fine - as long as the main check suite we have now passes, we should be confident merging - even if the artifacts take another 90 minutes to build from source.

I'm going to try and hook up as much of this as I can against my own fork and I'll submit a PR for that.

I'm really stunned github is giving away this much compute! If they increase the cache size from 2GB to ~50GB, these scratch builds could be a lot faster and cache all dependencies, only rebuilding bpftrace itself like the other builds do.

**Comment by danobi** [comment]
So the iovisor org is on some legacy plan b/c there's still a private repo. This prevents us from getting the free github actions builds (see https://github.com/iovisor/bpftrace/actions/new ):
> GitHub Actions workflows can't be executed on this repository.
>
> The billing plan for iovisor does not include access to GitHub Actions. Please contact the 
> organization owner or billing manager for questions about the current billing plan.

@4ast said he's willing to "downgrade" the iovisor org but we've got to be sure b/c it's a one way street.

If we're all on board with github actions, I can go contact github support and figure this out for sure.

**Comment by danobi** [comment]
When #1041 merges, we'll see for sure if github actions will run (b/c the PR checks in GH actions configs)

**Comment by dalehamel** [comment]
I just checked in quay.io and I cannot add the bpftrace repo to it, I'm guessing because I'm not an admin. @drzaeus77 seems to have the permissions in quay.io to do this though. Alternatively if someone else has admin on bpftrace I can add them to quay.io to connect the repo.

I'd like to add release steps to the github actions that will push this bpftrace image with minimal packaging to quay.io/iovisor/bpftrace:TAG, and have master pushes go to `latest`. We could use that and separate image tags for the builder/toolchain image to speed up builds a lot, as we currently always rebuild those at the start of each CI run.

**Comment by dalehamel** [comment]
@fbs closed #1088 which will help to support more portable bpftrace images.

Who in the iovisor org can add deploy secrets to the project? I think we need to add secrets for:

- quay.io (i have admin here for the kubect-trace project, but can't add the build hooks to the bpftrace repo, or add a bot secret for it to push containers to quay.io)
- dockerhub (probably makes sense to push here, as it will allow for just "docker pull" easier)
- Ubuntu snaps
- Any other secrets for pushing deploy artifacts

This will allow pushing tags to these various distribution centers, automating the release process further.

@brendangregg @ajor can either of you help to add these secrets to the repo to link up to these marketplaces?

**Comment by danobi** [comment]
The quay.io webhook is added now

**Comment by dalehamel** [comment]
I was able to get an image building on quay, but not a size optimized one. The build of llvm/clang with MinSizeRel times out. I think a better approach would be to work with the QUAY_TOKEN as a secret added to the repo, like in this comment https://github.com/iovisor/bpftrace/pull/1139#issuecomment-584848093

The size difference is pretty substantial, and we're already building the artifact on github actions rather than having to do the build on quay.io, we can just push the ~35 MB image layer to it after the builds are done.

**Comment by danobi** [comment]
I spent the morning fiddling with flatpak. TLDR: I couldn't get it to work.

This is the build manifest I ended up with: https://gist.github.com/danobi/4d5e0101ffe73f7b6d37b7291ab95a13

I originally tried to get flatpak-builder to download artifacts from GH actions on build, but unfortunately GH restricts artifact downloads to logged in users and bots with tokens (which I think would be hard to get checked into flathub). So I tried manually putting the binary into the directory tree which worked.

Unfortunately, I got stuck trying to run bpftrace:
```shell
$ flatpak-builder --run build-flatpak/ org.iovisor.Bpftrace.json help
bwrap: Can't mkdir parents for /run/host/usr: Permission denied
```
Seems other people have the issue but I can't seem to find a solution anywhere. Running with sudo didn't fix it either:
```shell
$ sudo flatpak-builder --run build-flatpak/ org.iovisor.Bpftrace.json help
[sudo] password for daniel:
bwrap: Unable to open lock file /run/user/0/.flatpak/2543822396/.ref: No such file or directory
```

### Other thoughts

Not sure if it's worth pursuing flatpak any further. Maybe if it gets easier to package. There's also the syscall filtering flatpak runtime does (that I think you can disable with `--allow=devel`). It's kinda ugly for users to run bpftrace with flatpak. Gotta do something like `flatpak run org.iovisor.Bpftrace --allow=devel` or something. Not sure if there's a way to put it in path.
